runtime/tchannel_server.go

```diff
@@ -64,197 +83,339 @@ func (ncr netContextRouter) Handle(ctx netContext.Context, call *tchannel.Inboun
+	if sep == -1 {
+		return "", "", false
+	}
+	return s[:sep], s[sep+2:], true
+}
+
 func (s *TChannelRouter) onError(err error) {
 	if tchannel.GetSystemErrorCode(err) == tchannel.ErrCodeTimeout {
-		s.logger.Warn("Thrift server timeout",
-			zap.String("error", err.Error()),
-		)
+		s.logger.Warn("Thrift server timeout", zap.Error(err))
 	} else {
-		s.logger.Error("Thrift server error.",
-			zap.String("error", err.Error()),
-		)
+		s.logger.Error("Thrift server error.", zap.Error(err))
 	}
 }
 
-func (s *TChannelRouter) handle(ctx context.Context, handler handler, service string, method string, call *tchannel.InboundCall) error {
-	treader, err := call.Arg2Reader()
-	if err != nil {
-		return errors.Wrapf(err, "could not create arg2reader for inbound call: %s::%s", service, method)
+func (s *TChannelRouter) handle(
+	ctx context.Context,
+	e *TChannelEndpoint,
+	call *tchannel.InboundCall,
+) (err error) {
+	c := tchannelInboundCall{
+		endpoint: e,
+		call:     call,
 	}
+	defer func() { c.finish(err) }()
+	c.start()
 
-	headers, err := ReadHeaders(treader)
-	if err != nil {
-		_ = treader.Close()
-
-		return errors.Wrapf(err, "could not reade headers for inbound call: %s::%s", service, method)
+	if err := c.readReqHeaders(); err != nil {
+		return err
 	}
-	if err := EnsureEmpty(treader, "reading request headers"); err != nil {
-		_ = treader.Close()
-
-		return errors.Wrapf(err, "could not ensure arg2reader is empty for inbound call: %s::%s", service, method)
+	wireValue, err := c.readReqBody()
+	if err != nil {
+		return err
 	}
 
-	if err := treader.Close(); err != nil {
-		return errors.Wrapf(err, "could not close arg2reader for inbound call: %s::%s", service, method)
-	}
+	// trace request
+	tracer := tchannel.TracerFromRegistrar(s.registrar)
+	ctx = tchannel.ExtractInboundSpan(ctx, call, c.reqHeaders, tracer)
 
-	treader, err = call.Arg3Reader()
+	resp, err := c.handle(ctx, &wireValue)
 	if err != nil {
-		return errors.Wrapf(err, "could not create arg3reader for inbound call: %s::%s", service, method)
+		return err
+	}
+	if err = c.writeResHeaders(); err != nil {
+		return err
+	}
+	if err = c.writeResBody(resp); err != nil {
+		return err
 	}
 
-	buf := GetBuffer()
-	defer PutBuffer(buf)
-	if _, err := buf.ReadFrom(treader); err != nil {
-		_ = treader.Close()
+	return err
+}
 
-		return errors.Wrapf(err, "could not read from arg3reader for inbound call: %s::%s", service, method)
-	}
+func (c *tchannelInboundCall) start() {
+	c.startTime = time.Now()
+}
 
-	tracer := tchannel.TracerFromRegistrar(s.registrar)
-	ctx = tchannel.ExtractInboundSpan(ctx, call, headers, tracer)
+func (c *tchannelInboundCall) finish(err error) {
+	c.finishTime = time.Now()
 
-	wireValue, err := protocol.Binary.Decode(bytes.NewReader(buf.Bytes()), wire.TStruct)
+	// emit metrics
 	if err != nil {
-		return errors.Wrapf(err, "could not decode arg3 for inbound call: %s::%s", service, method)
+		c.endpoint.metrics.SystemErrors.Inc(1)
```

Comment:
There are other system errors. We are going to really lose metrics here.

For example tchannel handles timeouts internally. I strongly suspects that if the method times out we do not capture this properly in `finish()`

Reply:
Looking into `ctx.Done()` to capture timeout errors.